### PR TITLE
Add new rules engine for section enabled rules

### PR DESCRIPTION
**Pull request by petechd**
### What is the context of this PR?
This adds new rules engine for the section enabled element of schemas, based on element type it will now evaluate `when` rules using old or new when rule evaluation. All schemas that use that used old type of when rules in section enabled has been updated and added to schemas with `new` prefix.

### How to review 
Run new test schemas manually or run unit tests.

### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

### CODE REVIEW

**Comment by petechd** [comment]
[Runner to use new rules engine for section enabled rules (3 SP)](https://trello.com/c/JKht4nws/4948-runner-to-use-new-rules-engine-for-section-enabled-rules-3-sp)

File: app/questionnaire/router.py [hunk 1]
```diff
@@ -270,15 +271,31 @@ def _is_section_enabled(self, section):
         if "enabled" not in section:
             return True
 
-        for condition in section["enabled"]:
-            if evaluate_when_rules(
-                condition["when"],
+        enabled = section["enabled"]
+        if isinstance(enabled, Sequence):
```
Comment (MebinAbraham) [inline app/questionnaire/router.py:275]: To be consistent with pathfinder, I think we should flip this logic. So check if it is a dict first then handle that, then handle list/tuple which is the old after.

File: app/questionnaire/router.py [hunk 2]
```diff
@@ -270,15 +271,31 @@ def _is_section_enabled(self, section):
...
                 self._list_store,
-            ):
-                return True
+                self._metadata,
+                self._response_metadata,
+                location=location,
+                routing_path_block_ids=None,
+            )
+
+            return when_rule_evaluator.evaluate(enabled.get("when"))
```
Comment (MebinAbraham) [inline app/questionnaire/router.py:297]: The `when` key has to exist so don't need `.get()`.

File: app/questionnaire/router.py [hunk 3]
```diff
@@ -270,15 +271,31 @@ def _is_section_enabled(self, section):
...
+                if evaluate_when_rules(
+                    condition["when"],
+                    self._schema,
+                    self._metadata,
+                    self._answer_store,
+                    self._list_store,
+                ):
+                    return True
+        else:
+            location = Location(section_id=section["id"])
```
Comment (MebinAbraham) [inline app/questionnaire/router.py:286]: I don't follow why this has been done?.

File: test_schemas/en/test_new_section_enabled_checkbox.json [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                {
+                    "when": [
+                        {
+                            "id": "section-1-answer",
+                            "condition": "contains",
+                            "value": "Section 3"
+                        }
+                    ]
+                }
+            ],
```
Comment (MebinAbraham) [inline test_schemas/en/test_new_section_enabled_checkbox.json:139]: This is the old rules structure.

**Review by MebinAbraham — COMMENTED** [review]
Not sure where we stand anymore with tests, but we need to update the references in test to the new schema or duplicate them as the new ones are being tested.

File: test_schemas/en/test_new_section_enabled_checkbox.json (hunk 4 above)
Comment (petechd) [reply test_schemas/en/test_new_section_enabled_checkbox.json]: Apologies, somehow it got into the new schema, fixed now, thanks!

File: app/questionnaire/router.py (hunk 2 above)
Comment (petechd) [reply app/questionnaire/router.py]: Removed unnecessary `.get()`, thanks!

File: app/questionnaire/router.py (hunk 1 above)
Comment (petechd) [reply app/questionnaire/router.py]: Logic flipped now, so first checking for dict and then `else`, thanks!

File: app/questionnaire/router.py (hunk 3 above)
Comment (petechd) [reply app/questionnaire/router.py]: So I removed it from here and made `location` optional in `WhenRuleEvaluator` but that lead to new line of code, assigning `list_item_id` a conditional value as it is part of Location class (and that is done in some other place in runner as well):
`list_item_id = self.location.list_item_id if self.location else None`

**Comment by petechd** [comment]
All the tests are referencing new schemas now except one that I had to keep in order to get 100% coverage (old evaluator engine code was missing tests).

File: app/questionnaire/routing/when_rule_evaluator.py [hunk 5]
```diff
@@ -19,19 +19,20 @@ class WhenRuleEvaluator:
     list_store: ListStore
     metadata: Mapping
     response_metadata: Mapping
-    location: Union[Location, RelationshipLocation]
+    location: Optional[Union[Location, RelationshipLocation]]
```
Comment (MebinAbraham) [inline app/questionnaire/routing/when_rule_evaluator.py:22]: As per the style guide, optional when we have one type. 
```suggestion
    location: Union[None, Location, RelationshipLocation]
```

**Review by MebinAbraham — APPROVED** [review]

File: app/questionnaire/router.py [hunk 6]
```diff
@@ -270,16 +271,30 @@ def _is_section_enabled(self, section):
...
                 self._schema,
                 self._metadata,
                 self._answer_store,
                 self._list_store,
-            ):
-                return True
-        return False
+            )
+            for condition in enabled
+        )
```
Comment (MebinAbraham) [inline app/questionnaire/router.py:297]: Nice use of `any`.

Comment (petechd) [reply app/questionnaire/router.py:297]: I was not sure if you'll prefer simplicity over readability here but "simple is better than complex".

**Review by LJBabbage — APPROVED** [review]
